File: Java/src/graph/BreadthFirstSearch.java
package graph;

import base.Queue;

/**
 * 图的广度优先搜索
 * 
 * @author igaozp
 * @since 2017-07-12
 * @version 1.0
 */
public class BreadthFirstSearch {
    /**
     * 存储图顶点的访问情况
     */
    private boolean[] marked;
    /**
     * 存储顶点的访问次数
     */
    private int count;

    /**
     * 够着方法
     * 
     * @param G 初始化的图
     * @param s 搜索的起点
     */
    public BreadthFirstSearch(Graph G, int s) {
        marked = new boolean[G.V()];
        bfs(G, s);
    }

    /**
     * 广度优先搜索
     * 
     * @param G 需要搜索的图
     * @param s 搜索的起点
     */
    private void bfs(Graph G, int s) {
        // 将访问过的元素加入队列
        Queue<Integer> queue = new Queue<>();
        marked[s] = true;
        queue.enqueue(s);

        // 依次访问以出队的元素为顶点的元素
        while (!queue.isEmpty()) {
            int v = queue.dequeue();
            for (int w : G.adj(v)) {
                // 没有访问则入队
                if (!marked[w]) {
                    marked[w] = true;
                    queue.enqueue(w);
                }
            }
        }
    }

    /**
     * 检查指定的顶点是否访问过
     * 
     * @param w 指定的顶点
     * @return {@code true} 访问过
     *         {@code false} 未访问过
     */
    public boolean marked(int w) {
        return marked[w];
    }

    /**
     * 获取图的访问次数
     * 
     * @return 访问次数
     */
    public int count() {
        return count;
    }
}

File: Kotlin/src/io/metatom/graph/BreadthFirstSearch.kt
package io.metatom.graph

import edu.princeton.cs.algs4.Graph
import edu.princeton.cs.algs4.Queue

/**
 * 图的广度优先搜索
 *
 * @author igaozp
 * @since 2017-09-08
 * @version 1.0
 */
class BreadthFirstSearch(G: Graph, s: Int) {
    // 存储图顶点的访问情况
    private var marked: MutableList<Boolean> = MutableList(0) { false }
    // 存储顶点的访问次数
    private var count: Int = 0

    // 构造方法
    init {
        marked = MutableList(G.V()) { false }
        bfs(G, s)
    }

    // 广度优先搜索
    private fun bfs(G: Graph, s: Int) {
        // 将访问过的元素加入队列
        val queue = Queue<Int>()
        marked[s] = true
        queue.enqueue(s)

        // 依次访问以出队的元素为顶点的元素
        while (!queue.isEmpty) {
            val v = queue.dequeue()
            for (w in G.adj(v)) {
                // 没有访问则入队
                if (!marked[w]) {
                    marked[w] = true
                    queue.enqueue(w)
                }
            }
        }
    }

    // 检查指定的顶点是否访问过
    fun marked(w: Int): Boolean = marked[w]

    // 获取图的访问次数
    fun count(): Int = count
}